File: graphs/UndirectedGraph.java
import java.util.*;

// Implementation of an undirected graph data structure using adjacency list representation with HashSet

public class UndirectedGraph
{
    private int V;
    private int E;
    private HashSet<Integer>[] adjList;

    public UndirectedGraph(int V)
    {
        this.V = V;
        E = 0;
        adjList = (HashSet<Integer>[]) new HashSet[V];

        for (int i = 0; i < V; ++i)
            adjList[i] = new HashSet<Integer>();
    }

    public int V() { return V; }
    public int E() { return E; }

    public void addEdge(int v, int w)
    {
        adjList[v].add(w);
        adjList[w].add(v);
    }

    public Iterable<Integer> adj(int V)
    {
        return adjList[V];
    }
}

File: graphs/UndirectedGraph.rb
require 'set'

class UndirectedGraph
	attr_reader :V, :E

	def initialize(num_vertices)
		@adjList = Array.new(num_vertices) { |i| i = Set.new }
		@V = num_vertices
		@E = 0
	end

	def addEdge(v, w)
		@adjList[v].add(w)
		@adjList[w].add(v)
	end

	def adj(v)
		return @adjList[v]
	end
end